boa_engine/src/builtins/array_buffer/mod.rs

```diff
@@ -389,39 +389,51 @@ impl ArrayBuffer {
         &self,
         src_byte_offset: u64,
         src_length: u64,
-        clone_constructor: &JsValue,
         context: &mut Context<'_>,
     ) -> JsResult<JsObject> {
-        // 1. Let targetBuffer be ? AllocateArrayBuffer(cloneConstructor, srcLength).
-        let target_buffer = Self::allocate(clone_constructor, src_length, context)?;
+        // 1. Assert: IsDetachedBuffer(srcBuffer) is false.
+        debug_assert!(!self.is_detached_buffer());
+
+        // 2. Let targetBuffer be ? AllocateArrayBuffer(%ArrayBuffer%, srcLength).
+        let target_buffer = Self::allocate(
+            &context
+                .realm()
+                .intrinsics()
+                .constructors()
+                .array_buffer()
+                .constructor()
+                .into(),
+            src_length,
+            context,
+        )?;
 
-        // 2. If IsDetachedBuffer(srcBuffer) is true, throw a TypeError exception.
         // 3. Let srcBlock be srcBuffer.[[ArrayBufferData]].
-        let src_block = self.array_buffer_data.as_deref().ok_or_else(|| {
-            JsNativeError::syntax().with_message("Cannot clone detached array buffer")
-        })?;
-
-        {
-            // 4. Let targetBlock be targetBuffer.[[ArrayBufferData]].
-            let mut target_buffer_mut = target_buffer.borrow_mut();
-            let target_block = target_buffer_mut
-                .as_array_buffer_mut()
-                .expect("This must be an ArrayBuffer");
-
-            // 5. Perform CopyDataBlockBytes(targetBlock, 0, srcBlock, srcByteOffset, srcLength).
-            copy_data_block_bytes(
-                target_block
-                    .array_buffer_data
-                    .as_mut()
-                    .expect("ArrayBuffer cannot me detached here"),
-                0,
-                src_block,
-                src_byte_offset as usize,
-                src_length as usize,
-            );
-        }
+        let src_block = self
+            .array_buffer_data
+            .as_deref()
+            .expect("ArrayBuffer cannot be detached");
+
+        // 4. Let targetBlock be targetBuffer.[[ArrayBufferData]].
+        let mut target_buffer_mut = target_buffer.borrow_mut();
+        let target_array_buffer = target_buffer_mut
+            .as_array_buffer_mut()
+            .expect("This must be an ArrayBuffer");
+        let target_block = target_array_buffer
+            .array_buffer_data
+            .as_mut()
+            .expect("ArrayBuffer cannot me detached here");
+
+        // 5. Perform CopyDataBlockBytes(targetBlock, 0, srcBlock, srcByteOffset, srcLength).
+        copy_data_block_bytes(
+            target_block,
+            0,
+            src_block,
+            src_byte_offset as usize,
+            src_length as usize,
+        );
 
         // 6. Return targetBuffer.
+        drop(target_buffer_mut);
```

Comment:
do we need to `drop` here since it will drop anyway at the end of the function.

Reply:
I think yes if we do not want to clone `target_buffer` on return. As far as I see it `target_buffer_mut` borrows from `target_buffer` and the drop at the end of the function happens after the move for the retun of the value. For example this would be the error if we remove the drop:

```
error[E0505]: cannot move out of `target_buffer` because it is borrowed
   --> boa_engine/src/builtins/array_buffer/mod.rs:437:12
    |
398 |         let target_buffer = Self::allocate(
    |             ------------- binding `target_buffer` declared here
...
417 |         let mut target_buffer_mut = target_buffer.borrow_mut();
    |                                     -------------------------- borrow of `target_buffer` occurs here
...
437 |         Ok(target_buffer)
    |            ^^^^^^^^^^^^^ move out of `target_buffer` occurs here
438 |     }
    |     - borrow might be used here, when `target_buffer_mut` is dropped and runs the `Drop` code for type `GcRefMut`
```